upb_generator/protoc-gen-upb.cc

```diff
@@ -888,13 +889,16 @@ void WriteHeader(const DefPoolPair& pools, upb::FileDefPtr file,
 
   if (options.bootstrap) {
     for (auto message : this_file_messages) {
-      output("extern const upb_MiniTable* $0();\n", MessageInitName(message));
+      output("extern const$1 upb_MiniTable* $0();\n", MessageInitName(message),
```

Comment:
The "bootstrap" output is only used during bootstrapping and should never be built as part of a DLL. It should only be linked into the `stage0` code generators themselves, which should only be statically linked.

Reply:
It looks like this comment has not been addressed. Nothing in the bootstrap stage should be affected by `dllexport_decl`.

Reply:
Sorry missed this - looking now to see...

Reply:
@haberman, a comment from the peanut gallery here[^1]

> [...] should never be built as part of a DLL.

Building grpc as a shared library did run into exactly these missing symbols though... As an example (taken from 1.57), it required:
```diff
diff --git a/src/core/ext/upb-generated/src/proto/grpc/gcp/altscontext.upb.h b/src/core/ext/upb-generated/src/proto/grpc/gcp/altscontext.upb.h
index 28c2185afd..68b940b360 100644
--- a/src/core/ext/upb-generated/src/proto/grpc/gcp/altscontext.upb.h
+++ b/src/core/ext/upb-generated/src/proto/grpc/gcp/altscontext.upb.h
@@ -27,7 +27,7 @@ extern "C" {
 
 typedef struct grpc_gcp_AltsContext grpc_gcp_AltsContext;
 typedef struct grpc_gcp_AltsContext_PeerAttributesEntry grpc_gcp_AltsContext_PeerAttributesEntry;
-extern const upb_MiniTable grpc_gcp_AltsContext_msg_init;
+extern const GRPC_DLL upb_MiniTable grpc_gcp_AltsContext_msg_init;
 extern const upb_MiniTable grpc_gcp_AltsContext_PeerAttributesEntry_msg_init;
 struct grpc_gcp_RpcProtocolVersions;
 extern const upb_MiniTable grpc_gcp_RpcProtocolVersions_msg_init;
diff --git a/src/core/ext/upb-generated/src/proto/grpc/health/v1/health.upb.h b/src/core/ext/upb-generated/src/proto/grpc/health/v1/health.upb.h
index 9217ea2312..e8a4b2b0c5 100644
--- a/src/core/ext/upb-generated/src/proto/grpc/health/v1/health.upb.h
+++ b/src/core/ext/upb-generated/src/proto/grpc/health/v1/health.upb.h
@@ -27,8 +27,8 @@ extern "C" {
 
 typedef struct grpc_health_v1_HealthCheckRequest grpc_health_v1_HealthCheckRequest;
 typedef struct grpc_health_v1_HealthCheckResponse grpc_health_v1_HealthCheckResponse;
-extern const upb_MiniTable grpc_health_v1_HealthCheckRequest_msg_init;
-extern const upb_MiniTable grpc_health_v1_HealthCheckResponse_msg_init;
+extern const GRPC_DLL upb_MiniTable grpc_health_v1_HealthCheckRequest_msg_init;
+extern const GRPC_DLL upb_MiniTable grpc_health_v1_HealthCheckResponse_msg_init;
 
 typedef enum {
   grpc_health_v1_HealthCheckResponse_UNKNOWN = 0,
diff --git a/src/core/ext/upb-generated/xds/data/orca/v3/orca_load_report.upb.h b/src/core/ext/upb-generated/xds/data/orca/v3/orca_load_report.upb.h
index 163152c592..3fb692e896 100644
--- a/src/core/ext/upb-generated/xds/data/orca/v3/orca_load_report.upb.h
+++ b/src/core/ext/upb-generated/xds/data/orca/v3/orca_load_report.upb.h
@@ -29,7 +29,7 @@ typedef struct xds_data_orca_v3_OrcaLoadReport xds_data_orca_v3_OrcaLoadReport;
 typedef struct xds_data_orca_v3_OrcaLoadReport_RequestCostEntry xds_data_orca_v3_OrcaLoadReport_RequestCostEntry;
 typedef struct xds_data_orca_v3_OrcaLoadReport_UtilizationEntry xds_data_orca_v3_OrcaLoadReport_UtilizationEntry;
 typedef struct xds_data_orca_v3_OrcaLoadReport_NamedMetricsEntry xds_data_orca_v3_OrcaLoadReport_NamedMetricsEntry;
-extern const upb_MiniTable xds_data_orca_v3_OrcaLoadReport_msg_init;
+extern const GRPC_DLL upb_MiniTable xds_data_orca_v3_OrcaLoadReport_msg_init;
 extern const upb_MiniTable xds_data_orca_v3_OrcaLoadReport_RequestCostEntry_msg_init;
 extern const upb_MiniTable xds_data_orca_v3_OrcaLoadReport_UtilizationEntry_msg_init;
 extern const upb_MiniTable xds_data_orca_v3_OrcaLoadReport_NamedMetricsEntry_msg_init;
```

So not all `upb_MiniTable` were affected (see surrounding lines), but it does at least seem to happen quite naturally that these symbols end up in the final DLL. The affected files correspond to the [workaround](https://github.com/grpc/grpc/blob/v1.59.0/CMakeLists.txt#L430-L444) Dawid had to introduce in the initial support for shared builds in grpc.

Note that at the time, this was happening in the checked-in, upb-generated sources. For [grpc 1.61](https://github.com/grpc/grpc/issues/35794), it wasn't necessary to patch stuff in [`upb-gen`](https://github.com/grpc/grpc/tree/master/src/core/ext/upb-gen), but rather I needed to patch the grpc-vendored upb itself (again for `upb_MiniTable` instances) to get shared builds to compile at all:
```diff
diff --git a/third_party/upb/upb/mini_table/internal/message.c b/third_party/upb/upb/mini_table/internal/message.c
index b25edacca7..05da188fc7 100644
--- a/third_party/upb/upb/mini_table/internal/message.c
+++ b/third_party/upb/upb/mini_table/internal/message.c
@@ -7,7 +7,7 @@
 
 #include "upb/mini_table/internal/message.h"
 
-const struct upb_MiniTable _kUpb_MiniTable_Empty = {
+UPB_DLL const struct upb_MiniTable _kUpb_MiniTable_Empty = {
     .subs = NULL,
     .fields = NULL,
     .size = 0,
diff --git a/third_party/upb/upb/mini_table/internal/message.h b/third_party/upb/upb/mini_table/internal/message.h
index 3457ea2cd6..8fa3325911 100644
--- a/third_party/upb/upb/mini_table/internal/message.h
+++ b/third_party/upb/upb/mini_table/internal/message.h
@@ -65,7 +65,7 @@ extern "C" {
 #endif
 
 // A MiniTable for an empty message, used for unlinked sub-messages.
-extern const struct upb_MiniTable _kUpb_MiniTable_Empty;
+UPB_DLL extern const struct upb_MiniTable _kUpb_MiniTable_Empty;
 
 // Computes a bitmask in which the |l->required_count| lowest bits are set,
 // except that we skip the lowest bit (because upb never uses hasbit 0).
```

If I understand you correctly, you seem to be saying that this is a bug in the way grpc does shared windows builds? 
Seeing that this always seems to involve `upb_MiniTable`, perhaps it's "too easy" to make that mistake though? (not that very many people are doing shared builds on windows...)

[^1]: I help maintain protobuf&grpc builds in the conda-forge distribution, we've been doing shared grpc builds on windows since 1.55, which was the jumping off point for the work that @dawidcha has been driving forward

Reply:
> As an example (taken from 1.57), it required:

All of the symbols you quoted (`extern const struct`) should indeed get the DLL export macro, I agree. My comment was not about them.

My comment was about a different set of generated files, a special and very specific kind of generated code that is only used when building upb itself and its core internal protos `descriptor.proto` and `plugin.proto`.

Because these protos are used internally by upb, these protos have to go through a special bootstrapping process when building upb from scratch in the GitHub repo. The bootstrapped version of these protos does not use `extern const struct`; it actually emits *functions* rather that const structs for the MiniTables.  You can see this in the very line we are discussing on this comment thread: https://github.com/protocolbuffers/protobuf/blob/fe6edca15b2d80df79ab31117dc49621ef461202/upb_generator/protoc-gen-upb.cc#L892

You can also see what the bootstrap generated code looks like, as we check it in to the source repo: https://github.com/protocolbuffers/protobuf/blob/fe6edca15b2d80df79ab31117dc49621ef461202/upb/reflection/stage0/google/protobuf/descriptor.upb.h#L20-L68

As you can see, this is clearly a different set of generated code than the samples you quoted, which should indeed get the macro, I agree.

It should be very easy to resolve my comment on this thread by simply reverting the change on this line, while still providing the export macros needed by end users.

Reply:
Thanks a lot for the response!

> As you can see, this is clearly a different set of generated code than the samples you quoted, which should indeed get the macro, I agree.

Yeah, I'm not surprised that there are other ways where `upb_MiniTable` things get generated, even though I had tried to double-checked whether I had missed other occurrences in `protoc-gen-upb.cc`. Wonderful news if it's just a one-line revert that's necessary!

Reply:
> It should be very easy to resolve my comment on this thread by simply reverting the change on this line, while still providing the export macros needed by end users.

@dawidcha, are you still planning to work on this PR? Otherwise I can try my hand at carrying this effort into another PR (or you invite me to your fork so I can push into this PR directly).

Reply:
@h-vetinari It would be great if you could finish this in a new PR!

@dawidcha Thanks for the awesome work so far!

Reply:
I've continued this PR in https://github.com/protocolbuffers/protobuf/pull/17079 and removed the changes to the bootstrap case.